### PR TITLE
Pi build and initial ArmCompute library support

### PR DESCRIPTION
**Pull request by quickwritereader**
## What changes were proposed in this pull request?

Pi build and initial ArmCompute library support
- Maxpool2d
- Avgpool2d

## How was this patch tested?

unit tests

## Quick checklist

The following checklist helps ensure your PR is complete:

- [x] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [x] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [x] Created tests for any significant new code additions.
- [x] Relevant tests for your changes are passing.

### CODE REVIEW

**Comment by saudet** [comment]
BTW, MKL-DNN/DNNL/oneDNN now support ARM:
> oneDNN CPU engine is used to execute primitives on Intel Architecture Processors, 64-bit Arm Architecture (AArch64) processors, and compatible devices.

https://github.com/oneapi-src/oneDNN/

**Comment by raver119** [comment]
Ye, it says 

```


    WARNING

    Arm 64-bit Architecture (AArch64) support is experimental with limited testing validation.
```

So one day it'll become a viable option we hope.

File: libnd4j/cmake/FindARMCOMPUTE.cmake [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+################################################################################
+
+message("<FindARMCOMPUTE.cmake>")
```
Comment (raver119) [inline libnd4j/cmake/FindARMCOMPUTE.cmake]: Do we really need this message?

File: libnd4j/cmake/FindARMCOMPUTE.cmake [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    
+    set(ARMCOMPUTE_LIBRARIES  ${ARMCOMPUTE_LIBRARY} ${ARMCOMPUTE_CORE_LIBRARY} )
+endif()
+ 
+ 
+INCLUDE(FindPackageHandleStandardArgs)
+
+FIND_PACKAGE_HANDLE_STANDARD_ARGS(ARMCOMPUTE REQUIRED_VARS ARMCOMPUTE_INCLUDE ARMCOMPUTE_LIBRARIES)
+
+message("</FindARMCOMPUTE.cmake>")
```
Comment (raver119) [inline libnd4j/cmake/FindARMCOMPUTE.cmake]: Same question :)

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+     }; 
+
+    return ret;
+}
+bool isArmcomputeFriendly(const NDArray& arr) {
+  auto dType = getArmType(arr.dataType());
+  int rank = (int)(arr.rankOf());
+  auto arrStrides = arr.stridesOf();
+  return dType != Arm_DataType::UNKNOWN && arr.ordering() == 'c' &&
+         arrStrides[rank - 1] == 1;
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: > arrStrides[rank - 1] == 1;

As far as I understand you want to make sure array isn't permuted. That's wrong check in this case.
You should check for ews being equal to 1 AND shape::strideDescendingCAscendingF() == true

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+Arm_TensorInfo getArmTensorInfo(int rank, Nd4jLong* bases,sd::DataType ndArrayType, arm_compute::DataLayout layout) {
+    constexpr int numChannels = 1; 
+    auto dType = getArmType(ndArrayType);
+    if (dType == Arm_DataType::UNKNOWN) {
+        //should not happen when isArmcomputeFriendly check is used
+        throw std::runtime_error("unsupported Data Type");
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: Please use sd::datatype_exception::build(...) instead

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    if (rank > arm_compute::MAX_DIMS) {
+        throw std::runtime_error("the rank of the array is higher");
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: Please add rank value to exception.

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+  //
+  constexpr int numChannels = 1;
+  int rank = (int)(arr.rankOf());
+  auto bases = arr.shapeOf();
+  auto arrStrides = arr.stridesOf();
+  if (rank > arm_compute::MAX_DIMS) {
+    throw std::runtime_error("the rank of the array is higher");
+  }
+  if (arr.ordering() != 'c' || arrStrides[rank - 1] != 1) {
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: Same as above

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+            {
+                auto src = tensor_it.ptr(); 
+                memcpy(dest, src, copySize);
+                dest += copySize;
+            },
+            tensor_it);
+    }
+    else {
+        Nd4jLong coords[MAX_RANK] = {};
+        if(strides[rank-1]!=1){
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: Same as above. It's possible to have permuted array with last stride being equal to 1.

File: libnd4j/include/ops/declarable/platform/armcompute/maxpooling2d.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+//////////////////////////////////////////////////////////////////////////
+PLATFORM_CHECK(maxpool2d, ENGINE_CPU) { 
+    auto input = INPUT_VARIABLE(0);
+    auto output = OUTPUT_VARIABLE(0);
+    const int dH = INT_ARG(6);
+    const int dW = INT_ARG(7);
+    // Data types supported: QASYMM8/QASYMM8_SIGNED/F16/F32
+    auto dTypeInput = getArmType(input->dataType());
+    auto dTypeOutput = getArmType(output->dataType());  
+    bool is_supported = dH==1 && dW==1 && isArmcomputeFriendly(*input) && isArmcomputeFriendly(*output)
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/maxpooling2d.cpp:96]: Are we sure pooling supports any combination of Kernel sizes?

File: libnd4j/tests_cpu/layers_tests/ConvolutionTests2.cpp [hunk 9]
```diff
@@ -1113,7 +1113,8 @@ TYPED_TEST(TypedConvolutionTests2, maxpool2d_6) {
     ASSERT_EQ(ND4J_STATUS_OK, result.status());
 
     auto z = result.at(0);
-
+    exp.printIndexedBuffer("Expected");
+    z->printIndexedBuffer("Z");
```
Comment (raver119) [inline libnd4j/tests_cpu/layers_tests/ConvolutionTests2.cpp:1118]: Please comment out all prints in tests before merge.

**Review by raver119 — CHANGES_REQUESTED** [review]
Generally LGTM, but some tweaks to be done

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp (hunk 3 above)
Comment (quickwritereader) [reply libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: C order and last rank strides being 1 were for the purpose of importing memory possible. but now as we have copy options of ndArray  we can make it redundant and add subarray copy for that case too.

Comment (raver119) [reply libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: Sure, I'm just saying that you must ensure array isn't permuted. In such case this check will be insufficient.

Comment (quickwritereader) [reply libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: for now these cases dismissed with that check:
```
a = a.subarray({ NDIndex::all(), NDIndex::all(),NDIndex::all(), NDIndex::all(), NDIndex::point(0) }); 
a.reshapei({ batch_k1,batch_k2, M,K },a_order);
```

File: libnd4j/cmake/FindARMCOMPUTE.cmake (hunk 2 above)
Comment (quickwritereader) [reply libnd4j/cmake/FindARMCOMPUTE.cmake]: I'll remove them

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp (hunk 6 above)
Comment (quickwritereader) [reply libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp]: well. tbh my conclusion of the last stride being 1 could be faulty. but I believe even for NHWC it should be so.

File: libnd4j/include/ops/declarable/platform/armcompute/maxpooling2d.cpp (hunk 8 above)
Comment (quickwritereader) [reply libnd4j/include/ops/declarable/platform/armcompute/maxpooling2d.cpp:96]: I should check their source code

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    //             memcpy(dest, src, width * element_size);
+    //             offset = sd::inc_coords(bases, strides, coords, offset, rank, 1);
+    //         },
+    //         tensor_it);
+    // }
+}
+
+void copyToTensor(const NDArray& input, Arm_Tensor& outTensor) {
+    //only for C order
+    if (input.ordering() != 'c') return;
```
Comment (raver119) [inline libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp:209]: Maybe throw an exception here?

**Review by raver119 — APPROVED** [review]
LGTM

**Comment by raver119** [comment]
@sshepel we need two things:

1. Add ARM Compute Library into CI docker image
2. Try CI with -h armcompute switch, that'll enable ARM CL helpers

File: libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp (hunk 10 above)
Comment (quickwritereader) [reply libnd4j/include/ops/declarable/platform/armcompute/armcomputeUtils.cpp:209]: noted it. I will visit these functions again as armCompute lib support is on progress.

**Pull request merged**